### ISSUE TITLE
[Feature Request]: "pad to largest resolution" and "automatically switch to game's native resolutions"

### ISSUE THREAD

**Issue by Patrickdroid** [libretro/pcsx_rearmed#203]
Problem:

PCSX ReArmed stretches all resolutions to 320x240, but PSX games ran in a variety of resolutions. Whenever a PSX game switches resolutions to something other than 320x240, you get scaling artifacts even with integer scaling enabled.

Solution:

there are two ways to address this- we can automatically detect and switch resolutions to the game's native resolution whenever a switch occurs, OR we can pad each resolution to the largest resolution that the game uses. Both of these options would cause the aspect ratio to change during gameplay, but scaling artifacts would be completely eliminated at all times.

Using the example of FFVII on the PSX:

on overworld screens and some minigames, the resolution switches to 320x224. On menu screens the resolution switches to 364x240.

If we "pad to largest resolution," we have a couple options. Assuming a user-selected integer scale of 5x5 and a custom aspect ratio of 1600x1200, 364x240 would become 1456x960, with black bars on the sides and top/bottom: (72 + 1456 + 72)x(120 + 960 + 120).

OR, if the custom aspect ratio is set to 1820x1200, then 320x240 becomes (110 + 1600 + 110)x1200 and 364x240 becomes 1820x1200. So every resolution winds up fitting within the largest resolution at 5x5 integer.

With "automatically switch resolution to game's detected native resolution," you would simply take the game's currently detected native resolution and then scale it up by the user-selected integer scale factor. Returning the to example of FFVII, if we assume a user-selected integer scale factor of 5x5, we would get the following resolutions:

1600x1200 during normal gameplay (320x240)
1820x1200 on menu screens (364x240)
1600x1120 on the overworld screen and certain minigames (320x224)

It would be nice to have both options available: "pad to largest resolution" and "automatically switch to game's current native resolution," so that the user had more control over the output, allowing them to choose between taking up more of the display area or stretching the image less. Both options would result in pixel-perfect video output at all times.

**Comment by Sakitoshi** [comment]
I stated something similar in #180 
and while is not a solution you can work around the issue using one of the box shaders for the time being.

**Comment by jayare5** [comment]
> I stated something similar in #180
> and while is not a solution you can work around the issue using one of the box shaders for the time being.

Can you explain this about box shaders?
I have a problem with some games where the resolution will stretch too. But for me, it is terrible, because I'm using the CRT 240p mod on Raspberry Pi.
If you tried the PS1 version of Chrono Trigger, I think every map even uses a different resolution or something, and its really messy. Another game, Alundra 2, that's the most noticeable because it has a loading screen that is supposed to only occupy a small space in the corner, but because of the stretch problem, it fills up like almost the whole screen.

**Comment by Sakitoshi** [comment]
@jayare5 if you are using a raspi on a crt then you would be better off using my configs. I investigated the issue precisely because of that.
https://github.com/Sakitoshi/retropie-crt-tvout

**Comment by jayare5** [comment]
> @jayare5 if you are using a raspi on a crt then you would be better off using my configs. I investigated the issue precisely because of that.
> https://github.com/Sakitoshi/retropie-crt-tvout

Oh man, that sounds like it would be great!
However, I do not use retropie. I use Lakka.
It seems that the main thing at work here is the shader combined with 640x480 settings?
I tried to use the shaders with the glslp, all seemingly in the correct folders and all, but it did not even load it :/

**Comment by Sakitoshi** [comment]
@jayare5 I suggest you to switch to retropie if you plan on using your raspi on a crt tv, each system has his own config so it's much easier to make them look as it should. but I think you can still adapt my configs to lakka if you setup overrides for every system.

in the case of playstation, yes. 640x480 along with the tvout_smart_align shader preset does the work.
just copy the tvout folder inside the lakka shaders folder and it should work without needing any modifications.

**Comment by jayare5** [comment]
> @jayare5 I suggest you to switch to retropie if you plan on using your raspi on a crt tv, each system has his own config so it's much easier to make them look as it should. but I think you can still adapt my configs to lakka if you setup overrides for every system.
> 
> in the case of playstation, yes. 640x480 along with the tvout_smart_align shader preset does the work.
> just copy the tvout folder inside the lakka shaders folder and it should work without needing any modifications.

Hahahahaha, I already spent nearly 2 years and counting, configuring every system for multiple uses too. I am way too deep into Lakka to switch. 
All right, Im gonna try it again.

**Comment by jayare5** [comment]
> @jayare5 I suggest you to switch to retropie if you plan on using your raspi on a crt tv, each system has his own config so it's much easier to make them look as it should. but I think you can still adapt my configs to lakka if you setup overrides for every system.
> 
> in the case of playstation, yes. 640x480 along with the tvout_smart_align shader preset does the work.
> just copy the tvout folder inside the lakka shaders folder and it should work without needing any modifications.

Ok I have a question though. Must the shader be made compatible with every game manually by you?
Because I noticed a game that still have a scaling problem (Alundra 2's loading screen).
Can this be fixed?

**Comment by Sakitoshi** [comment]
@jayare5 the shader was modified by me, but I merely disabled the horizontal scaling of it.
ps1 games have a range of vertical resolutions, so adding them to a list would be too problematic, what the shader does is simply double the vertical resolution but only if the result is equal or less than the vertical viewport resolution you have set in retroarch, 480 in this case. as no game goes beyond 480 (talking about NTSC region here) all are scaled to 1x or 2x its vertical resolution leaving black bars to compensate (overscan).

if you are having problems I assume it's because you are using the raspi overscan setting instead of disabling it. the raspi overscan setting decreases the resolution to achieve its result, so you end up with a resolution of less than 480.
having said that I now realize that it doesn't really matter as the custom viewport setting of retroarch can be higher than the screen resolution, you'll just get a smaller viewable area.

**Comment by jayare5** [comment]
Hmmm... are you talking about about that Retroarch option that says "Crop Overscan (Reload)"?
It was disabled, but I enabled it to test, no difference.
This is what it looks like btw, it's a really low custom resolution the game uses in this particular screen, probably around 30 pixels height:
https://i.postimg.cc/v8L92Q2S/52991442-2118945044860686-2146385824294371328-n.jpg

**Comment by Sakitoshi** [comment]
@jayare5 no, I mean the /boot/config.txt file.
I've never seen the crop overscan option do anything at all tbh.

I went and downloaded a copy of the game to see for myself and looks like the emulator is failing to determine the real resolution of that screen. without any shader active also fails to show as expected.
the shader uses the "inputsize" varible which contains the emulator core resolution. the core does updates that variable, it just doesn't inform the retroarch gui, it tells it that 320x240 is the native resolution. so the shader is always in the know of what resolution is really being used.
there is nothing else I can do to help.

**Comment by jayare5** [comment]
Wow, that is unfortunate. Thank you very much for all your time and effort anyway!!! I really appreciate it. It's been many years and this emulator has not been fixed for that. A lot of other games did get fixed with the shader, I see Lunar SSSC playing movies at the correct resolution now, for example.
I doubt that any game would require gameplay at that 30 pixels height resolution anyway :P

**Comment by retro-wertz** [comment]
how does mdfn psx looks, basically this core and mdfn psx uses the same window size regardless of framebuffer size

**Comment by hizzlekizzle** [comment]
The core should now switch to match the game's actual resolution. Can this issue be closed?

**Comment by retro-wertz** [comment]
i really do not like this changing of resolution ingame, especially one like psx that changes a lot. for a monitor resolution of 900px height (Yeah, there is such a monitor). a 224 would fit perfectly in x3 size, but not in 240 so image will resize...

but...

**Comment by dankcushions** [comment]
i don't think it should resize the image. the reported aspect ratio is still 4:3. 

(unless you have integer scaling on, but you probably don't want it on in any scenario with a 900px high monitor)

**Comment by retro-wertz** [comment]
why not? every 224 height game (nes.snes) works fine with interger scaling.
scanline shader do not work well without interger scale. we should remove
interger scaling then if this is useless.

**Comment by dankcushions** [comment]
oh, fair, i got confused by your x3 comment.  224x4=896 px which would be a tiny border which i can agree you want.

buuuut, if the game is changing resolution behind the scenes then IMO you always want it to be reported to the API. if you don't, then your scanline shaders are going to be putting the wrong number of scanlines in the wrong places, and your integer scaling is no longer going to be working properly anyway - you will have scaling artefacts.

**Comment by jayare5** [comment]
In my opinion, the perfect PCSX ReArmed fix would have:

1. Locked minimum vertical resolution of either 224 or 240 (whatever it is the PS1 actually does)
2. Display the height at 1:1 what the game is doing.
3. Detect the correct coordinates of where to display the screen.
4. Horizontal res should have a "Core option", so that it can be either 1:1 or stretch:
1:1 would help greatly for those who want to take clean pixel perfect screenshots while playing or just simply prefer to view their games that way.
Stretch mode would help greatly to keep the aspect ratio of the original system.

The easiest screen to test is the Now Loading screen from Alundra 2.
https://www.youtube.com/watch?v=6mIYNgUrrVw
At 0:07, it appears that Now Loading is on the bottom right corner.
PCSX ReArmed currently takes that image and stretches it all over, and doesn't even seem to detect the coordinates properly, which is why the step 3 above is important.

**Comment by dankcushions** [comment]
but the psx doesn’t have a fixed resolution, it does many. often several resolutions in one game. that’s the issue being explored here. 

‘1:1’ is integer scaling in retroarch. enable it if you want it, but in my experience it’s rarely useful. ‘stretch mode’ is ‘core provided’ aspect ratio. it’s the default and is what i always want. in either case, they’re handled by the front end (retroarch). all the core needs to do is tell the front end what the current resolution and aspect ratio is, which it now does.

your alundra sounds like an emulation bug. unless you have the ‘enhanced resolution’ hack enabled, which causes various display issues in many games.

**Comment by jayare5** [comment]
Well, when you say "it now does" do you mean a recent update as of this month / last month? 

And I promise you, it's not an emulation bug :P I have no enhancements enabled. Every PS1 game does the same thing, just at different levels of intensity. Chrono Trigger PS1 has like a different resolution on each map! That game is a nightmare to make it look properly.
Most people dont notice this because they're playing on PC monitors or HD TVs, where the games will have lots of pixels to be able to stretch and not have a noticeable change, especially with smooth shaders enabled. And thats why I used Alundra 2, because there's no way you aren't gonna notice that.

I notice all this though, because I play PS1 from RetroArch on a Raspberry Pi displaying at 240p on a CRT. Here it's really easy to see every time PCSX ReArmed displays the wrong resolution because the CRT is what the PS1 was designed for to begin with.

So, if Chrono Trigger and Alundra 2 both display properly on a CRT, using the same config, then it's good now. If not, we need to look into this further.

**Comment by dankcushions** [comment]
> Well, when you say "it now does" do you mean a recent update as of this month / last month?

it's linked above. see https://github.com/libretro/pcsx_rearmed/pull/307

> I notice all this though, because I play PS1 from Lakka on a Raspberry Pi displaying at 240p on a CRT

oh, you're using a CRT - that's a different consideration, but it sounds like this fix along with using retroarch's CRT switch res feature would fix all your issues. please get the latest nightly and then ask on the forums for support on using switch res.  some monster threads on it: https://forums.libretro.com/t/retroarch-native-crt-support/10124

**Comment by jayare5** [comment]
> > Well, when you say "it now does" do you mean a recent update as of this month / last month?
> 
> it's linked above. see #307
> 
> > I notice all this though, because I play PS1 from Lakka on a Raspberry Pi displaying at 240p on a CRT
> 
> oh, you're using a CRT - that's a different consideration, but it sounds like this fix along with using retroarch's CRT switch res feature would fix all your issues. please get the latest nightly and then ask on the forums for support on using switch res. some monster threads on it: https://forums.libretro.com/t/retroarch-native-crt-support/10124

Ok, I've tested the latest nightlies. PCSX ReARMed has not improved. It still stretches everything vertically, and it shouldn't. I have compared it to the Beetle core and taken some screenshots.
Beetle does the vertical res like a real PS1 would.
![PS1 vertical res 1](https://user-images.githubusercontent.com/26237704/62015557-8d85e680-b17a-11e9-8550-3b43687033ad.jpg)
![PS1 vertical res 2](https://user-images.githubusercontent.com/26237704/62015560-98d91200-b17a-11e9-8597-6cae3c089af9.jpg)
![PS1 vertical res 3](https://user-images.githubusercontent.com/26237704/62015564-a098b680-b17a-11e9-871d-5f1de3c111b1.jpg)
![PS1 vertical res 4](https://user-images.githubusercontent.com/26237704/62015566-a68e9780-b17a-11e9-9470-79bb2af4ab1e.jpg)
![PS1 vertical res 5](https://user-images.githubusercontent.com/26237704/62015558-92e33100-b17a-11e9-9010-4beee96cf1eb.jpg)

**Comment by retro-wertz** [comment]
for chrono cross at least, the height should be 211px, it shouldnt stretch like that. check if you have overrides or manual aspect set. but since this is for crt, dont have a way to check that.
 
![Screenshot_2019-07-29_09-38-04](https://user-images.githubusercontent.com/23315382/62016792-acba5d80-b1e6-11e9-8bd7-74ecd9885927.png)

**Comment by Sakitoshi** [comment]
@retro-wertz even with the new auto-res he would not get what he wants.
I've been using pcsx rearmed on a crt from quite some time now (on retropie, but shouldn't matter) and the only solution is to use a fixed resolution and let shader center the picture vertically.
ps1 games can change resolution on the fly but integer scaling doesn't adjust automatically (and shouldn't in my opinion, it would be a mess). only shaders can make changes like that on the fly and depending the input.
to solve this a more aggressive change will be needed. disable completely the overscan cut so the core always outputs 240 or 480 vertically (or 288 and 576 in the case of pal games) with black bars up and down to fill the gaps. I'm using a shader that does exactly that and only that way ps1 games look perfect.

**Comment by jayare5** [comment]
The screenshots I took were with all the default options... if you have to change the video height manually, that's not gonna fix anything because this game can actually call multiple resolutions. (and how are you gonna fix every other game that calls a random vertical res?)

However, if you want to see if it still causes the problem, you really dont need a CRT... you can just view it on your computer screen if you select 1x Window mode, and Aspect Ratio set to 1:1 PAR 4:3

![CT vert](https://user-images.githubusercontent.com/26237704/62018954-d6de3200-b18a-11e9-9645-99d059ff02ba.png)

**Comment by retro-wertz** [comment]
the way you are describing this, is more of overscan nature. you can't just get rid of and change heigt and not scale the game image also else it wouldnt look correct ( as the case with you sample where in the forest its 211px and 224 on the overworld... the geometry itself is set at 240 unless you refresh your viewport(because of the addition of this update geometry each change))

**Comment by jayare5** [comment]
Not only is this an overscan issue, it also doesnt seem to know where to place the screen.
For example the Now Loading screen appears aligned to the right when using the Beetle core.
It looks centered in PCSX ReARMed.  I probably wouldn't mind centered if it helps keep the emulator faster, but the vertical overscan / stretching issue needs to be solved.

**Comment by Sakitoshi** [comment]
disregard alundra 2 loading screen, that seems to be a problem of its own.
pcsx rearmed isn't rendering that particular screen correctly, causing the distortion.

**Comment by jayare5** [comment]
> disregard alundra 2 loading screen, that seems to be a problem of its own.
> pcsx rearmed isn't rendering that particular screen correctly, causing the distortion.

Hey so wait a minute, what were the changes mentioned by @dankcushions that were supposed to be in the nightly??

**Comment by dankcushions** [comment]
> it's linked above. see #307

and i already said that the alundra issue looks like some other bug:

> your alundra (issue) sounds like an emulation bug.

**Comment by jayare5** [comment]
> > it's linked above. see #307
> 
> and i already said that the alundra issue looks like some other bug:
> 
> > your alundra (issue) sounds like an emulation bug.

...Did you see the other pictures I posted? It seems to me nothing has been fixed or changed in the latest nightly, so I wonder if that's supposed to happen or not as of today.
Every game will stretch vertically.
Is there a particular game that has been fixed vertically?
And is there a specific setting I must use for the changes in the link above to take effect?

**Comment by dankcushions** [comment]
tbh i think yours is more of a support issue. pcsx rearmed is not vertically stretched here using default settings. i get the same as https://github.com/libretro/pcsx_rearmed/issues/203#issuecomment-515817780

you should give your full .cfg settings and core option settings.

**Comment by jayare5** [comment]
> tbh i think yours is more of a support issue. pcsx rearmed is not vertically stretched here using default settings. i get the same as [#203 (comment)](https://github.com/libretro/pcsx_rearmed/issues/203#issuecomment-515817780)
> 
> you should give your full .cfg settings and core option settings.

The #203 (comment) explicitly says they're using a custom setting of 211 for the height.
Which is only gonna look good in that place in that game (or inside a house) but not in the overworld.

My tests were done with a fully default config, it was a newly downloaded nightly and I just downloaded the cores on that day, I didn't even disable the Bilinear filter in Video that I normally would to get sharp pixels.
Both Beetle and PCSX ReARMed were their default configs in my screenshots.

Can you show a screenshot of how Chrono Trigger looks in that forest?
As well as a screenshot of the config in that same place?

**Comment by retro-wertz** [comment]
im not using any custom settings, these are default. The 211 is what CT uses for its height in this forest region, changes to 224 on overworld e.g.

**Comment by jayare5** [comment]
> im not using any custom settings, these are default. The 211 is what CT uses for its height in this forest region, changes to 224 on overworld e.g.

So what youre saying is, the Video setting changed to 211 on its own??
Then I'm definitely using some kind of outdated nightly or something.
I thought I was supposed to go here:

https://buildbot.libretro.com/nightly/windows/x86_64/

And download the setup with the latest date?

**Comment by AnotherLife** [comment]
https://buildbot.libretro.com/nightly/windows/x86_64/latest/

What about that one?

**Comment by jayare5** [comment]
Ohh!!
Ok, I didn't go there. Will try it ASAP.

**Comment by jayare5** [comment]
Ok I tried. No, it doesn't fix anything.

This is gonna be a lengthy post. I tried Chrono Trigger, all default settings. It looks like this. 
![CT](https://user-images.githubusercontent.com/26237704/62332183-81f13300-b48c-11e9-9236-9aae257eb0c2.png)

This is incorrect, because its supposed to have black borders.
I went to the video settings and messed with some things and tried again. 
Aspect Ratio was Core Provided, I made it 1:1 PAR.
Clicked the Window menu and made it 2x.
It actually still has no black borders on the top and bottom as it should.
![CT2](https://user-images.githubusercontent.com/26237704/62332325-15c2ff00-b48d-11e9-8a5e-39ad52a2ec57.png)

So now I leave the forest and...
![CT3](https://user-images.githubusercontent.com/26237704/62332338-283d3880-b48d-11e9-99dd-93f8873cc14f.png)

Now we have black borders ON THE SIDES. This is absolutely wrong in every way.
At this point, the Aspect Ratio option is completely useless to attempt to make this game display properly, so I take matters into my own hands and attempt to configure it manually, and did what a PS1 core SHOULD expect me to do.

![CT4](https://user-images.githubusercontent.com/26237704/62332451-b9acaa80-b48d-11e9-8e87-2a7f582a2426.png)

This config right here is what should work with EVERY PS1 game, just like an NES or SNES core would when you put a similar config. I know the PS1 is different internally, but Beetle does work properly this way (Although Beetle recognizes 480 as the 2x resolution where this screenshot shows that 448 is the 2x resolution, but that would be OK too if PCSX ReARMed had been made to work properly with a 448 setting.)

Anyway, I walked into the forest, and... no black borders. It's still getting stretched. I check my config, and now 448 is no longer the 2x resolution.
![CT5](https://user-images.githubusercontent.com/26237704/62332827-178dc200-b48f-11e9-980e-8594458d3f45.png)

You people insist in ignoring Alundra 2, but I just have a funny feeling that it explains everything that is wrong with the emulator because it does recognize the 2x resolution properly, which means once more that this emulator is stretching things vertically when it shouldnt. Or rather, that it shouldnt treat every in-game resolution as a different RetroArch resolution like it currently does. Instead it should be like the Beetle core, to always be 480 or 448, and display things properly inside that.

![Alundra 2](https://user-images.githubusercontent.com/26237704/62332973-88cd7500-b48f-11e9-82aa-67a8c738efbd.png)

That is all this emulator needs to do to be fixed, almost like the shader Sakitoshi posted, except without the need for a shader.

I don't understand how difficult it could be to make PCSX ReARMed to have a constant vertical resolution of 480 and display things 1:1 (480i mode) or 1:2 (240p mode) inside those 480, and not be dependent on what the game does, but that's what needs to be done. 

But wait, there's more.

![Lunar](https://user-images.githubusercontent.com/26237704/62333176-6b4cdb00-b490-11e9-98bb-354281da69a5.png)

Lunar's FMV seems vertically fine! However, it is horizontally wrong, there's too much border. This is still the 640x448 setting. I find that what it's actually doing is just displaying the movie at 2x2 its original res, and the window size from CT (which was correct) messed it up. Now setting it to this option reveals that PCSX ReARMed still doesnt recognize the black borders on the top and bottom of this FMV.

![Lunar2](https://user-images.githubusercontent.com/26237704/62333261-c54da080-b490-11e9-83eb-465ad9575154.png)

Lastly, I tried this on my Raspberry Pi 3b+, and PCSX ReARMed behaves the same there too, except it plays the games at like 5 fps instead of fullspeed like it used to be.

**Comment by dankcushions** [comment]
@jayare5 firstly, you should never use PAR 1:1 it you want proper display. that fundamentally isn't what that setting does. TVs don't require 'square' pixels so by setting it to that you will mess up the aspect ratio in many/most scenarios. always leave it to core provided when investigating issues of this nature.

let's deal with one thing at a time. chrono trigger in the 'town' mode (not world map). with the settings at core provided, it's stretching a 256x211px image to 4:3. you're saying it should have a border? that would seem strange to me. the game is sending a 256x211 signal, and psx games are always 4:3 DAR. if it had a border it should send a 256x211(+whatever height the border(s) are). eg, if you play a PAL game with borders, the emitted resolution will include those borders, not exclude them.

paging @hizzlekizzle

**Comment by jayare5** [comment]
@dankcushions 
Forget about PAR 1:1, I was only testing that because Core Provided, the default option, was doing things wrong.
TL;DR,  skip to the end of my previous post, it will be futile unless PCSX ReARMed handled the screen resolution exactly how Beetle does.

--------------------------------------------------------------

You say I should use "Core Provided", but you did you see how it added black borders on the SIDES of the screen in the overworld?
And did you see the photos of Chrono Trigger running on a real PS1?
1. Chrono Trigger's forest has a black border and it's on the bottom.
2. The overworld has no black borders anywhere.
3. And Lunar's FMV has equally sized (or near equally) black borders all around it.

Every horizontal line of pixels is there untouched, and it fills the screen horizontally.
The only way to achieve this in emulation is if we treated it like Beetle core does, it has a 640x480 in the resolution retting at all times, no matter what the game says, but it will display things inside it, ONLY horizontally stretched and NOT vertically. Vertical resolutions should always, ALWAYS be 1:1  or 1:2.
(1:1 - For when a game is 480i, like the BIOS boot logo or Gran Turismo's titlescreen)
(1:2 - For when a game is 240p like Chrono Trigger, or even Gran Turismo when in the FMV or in-game)

Here's my next set of screenshots using Core Provided, as you suggested.
![SettingsA](https://user-images.githubusercontent.com/26237704/62375993-a04b4300-b50d-11e9-8a2f-ae9711d31153.png)

Screenshots: 
![SettingsA1](https://user-images.githubusercontent.com/26237704/62376014-ad683200-b50d-11e9-8b2a-5a3bb319de36.png)
![SettingsA2](https://user-images.githubusercontent.com/26237704/62376228-2cf60100-b50e-11e9-8e55-125c9b497099.png)

PCSX ReARMed:
horizontally: stretched, CORRECT
vertically: stretched, INCORRECT

Beetle :
horizontally: stretched, CORRECT
vertically: even pixels, CORRECT (the black border below is correct.)


Next I tried Integer scaling cause I assume you may be using that to prevent stretching in PCSX ReARMed? Here's a picture of the settings:
![SettingsB](https://user-images.githubusercontent.com/26237704/62376305-5dd63600-b50e-11e9-97e4-7d4ec4432b3f.png)

Now here are screenshots of what that does:
![SettingsB1](https://user-images.githubusercontent.com/26237704/62376337-6d557f00-b50e-11e9-8767-f052b0de06a9.png)
![SettingsB2](https://user-images.githubusercontent.com/26237704/62376341-6f1f4280-b50e-11e9-992e-a782c6a7b644.png)

PCSX ReARMed:
horizontally: even pixels, INCORRECT, now we have extra black borders on the sides that shouldn't be there! Both screens even have different amount of horizontal borders, that's totally incorrect.
It probably is what the PS1 was doing internally, but it's not what the player should be seeing.
vertically: even pixels, CORRECT, this is good even though it's centered when it should not be, it should have specifically more black border on the bottom than on top for the forest but thats ok. But because it's now horizontally incorrect, we haven't fixed the problem.

Beetle :
horizontally: stretched, CORRECT
vertically: even pixels, CORRECT (the black border below is correct.)

It seems that in all possible cases, Beetle wins, so the only way that PCSX ReARMed is if it handled all these screen resolutions exactly the way Beetle does.

**Comment by retro-wertz** [comment]
in beetle, why is stretching a 256-width CT game to a 320 width correct?

**Comment by jayare5** [comment]
> in beetle, why is stretching a 256-width CT game to a 320 width correct?

Sadly, the PS1 is quite a complicated system when it comes to video display.
(Did you know I've tried a core for almost every console currently in RetroArch and PCSX ReARMed is the only one I had issues with?)
The way the PS1 works will probably not be ideal with how RetroArch handles resolution numbers, and Beetle seems to be made to work around it.
It gives priority to the TV aspect ratio that the original PS1 had over TV signals, and that's the better behavior to display things how they should be.

If I would have things COMPLETELY my way, I'd have a core specific option making it can switch between Horizontally 1:1 or Horozontally stretched. 

I think that making it horizontally 1:1 would help in devices that have low screen sizes like 320x240.
But because I don't use OR have such devices, I'd feel a bit bad asking for a feature that's not even relevant to me. Another use such a feature would have is to take pixel perfect screenshots of games.

> so which one is "correct" in this screenshot? pcsx which centered the image? or beetle which image is more to the top?
The correct one is the Beetle one with the image more to the top.
Have you completely ignored my photo of my TV running from a real PS1? Here it is again just in case: 
![62016552-cc1e9f80-b180-11e9-9f2a-24c6e54ca1fc](https://user-images.githubusercontent.com/26237704/62382770-44d48180-b51c-11e9-9d8b-8c47f4054585.jpg)

**Comment by Sakitoshi** [comment]
can confirm that chrono trigger has a black bar at the bottom of the screen on real hardware (fat ps2 in my case). this is also true for the original game on snes btw.
at this point I'll be more inclined to say that chrono trigger has the same bug as alundra 2 where black portions of the screen are completely ignored and treated as overscan and thus cropped.

**Comment by jayare5** [comment]
Yeah! What it seems to me is that, if the PS1 is truly calling different custom resolutions of any size, it seems that it also has the ability to put it in a custom coordinate on the screen.

**Comment by dankcushions** [comment]
> It gives priority to the TV aspect ratio that the original PS1 had over TV signals, and that's the better behavior to display things how they should be.

this is actually exactly how retroarch works. resolution and aspect ratios are separate. you can have a 100x100 game displayed at 4:3 no problem, if that’s what the emulator asks. the aspect ratio doesn’t appear to be the problem here, but the reported resolution.

> at this point I'll be more inclined to say that chrono trigger has the same bug as alundra 2 where black portions of the screen are completely ignored and treated as overscan and thus cropped.

yeah i think this is exactly what’s happening. i wonder if it happens in PEOPS or just NEON graphics plugin?

**Comment by jayare5** [comment]
> this is actually exactly how retroarch works. resolution and aspect ratios are separate. you can have a 100x100 game displayed at 4:3 no problem, if that’s what the emulator asks. the aspect ratio doesn’t appear to be the problem here, but the reported resolution.

I already know this, I was answering "why is stretching a 256-width CT game to a 320 width correct?"
which makes it 4:3 even when you use integer and the game was stretched horizontally rather than 256x

PCSX ReARMed thing is, there are games like FF7 that can go beyond 320x240, like in this screenshot:
![a10332499706cf967c22edbd5fed43b8](https://user-images.githubusercontent.com/26237704/62412428-452e5480-b5d0-11e9-8a79-410d509c04c1.png)

Although this isn't captured from inside RetroArch (I just looked for it online) but this screenshot is 364x224... I don't know what maximum width the PS1 could do while still doing only 224.
Maybe the PS1 can do 400x224? I do not know. But that would be more or less what's required to make it display 1:1 pixels with its integer option for all games, and the 4:3 option should turn that into 320x224.
(stretch/compress horizontally, leave vertical res alone)
Yet a 256x224 game would be inside that 400x224 box, and should be 320x224 in 4:3  (while still being inside a 400x224 box!
When you consider how RetroArch works, I don't think I've seen something like that work that way.

>at this point I'll be more inclined to say that chrono trigger has the same bug as alundra 2 where black portions of the screen are completely ignored and treated as overscan and thus cropped

> yeah i think this is exactly what’s happening. i wonder if it happens in PEOPS or just NEON graphics plugin?

Anyway, just remember one thing, it's not just those games, it can affect every game, and every screen that calls for anything with black borders on the top and bottom.

**Comment by dankcushions** [comment]
> Maybe the PS1 can do 400x224? I do not know. But that would be more or less what's required to make it display 1:1 pixels with its integer option for all games, and the 4:3 option should turn that into 320x224.
(stretch/compress horizontally, leave vertical res alone)

no, 4:3 would be 300x224, although i believe retroarch will use the largest side and adjust from there (400x300, then), in windowed mode. you could easily check...

> FINALLY We understand its not just "Forget Alundra 2 because it has a bug" but rather, "PCSX ReARMed has a bug with black borders / overscan cropping"

@jayare5 don't keep being a jerk about this. people are trying to help. you're commenting on an unrelated issue so this should really be logged separately. your issue isn't to do with games that change resolution, but games with borders. this is clutter at this point.

**Comment by jayare5** [comment]
>your issue isn't to do with games that change resolution

But it is...
Look what the OP says:

>Problem:
PCSX ReArmed stretches all resolutions to 320x240, but PSX games ran in a variety of resolutions. Whenever a PSX game switches resolutions to something other than 320x240, you get scaling artifacts even with integer scaling enabled.

I can promise you, if the scaling issue I'm showing in the screenshots was fixed, the OP wouldn't have had that problem.

>don't keep being a jerk about this. people are trying to help.

Are you really going to blame me after seeing this problem since like 2016 and no one handling it properly to this day? (Problem has probably existed for years prior to that too)

**Comment by dankcushions** [comment]
> Are you really going to blame me after seeing this problem since like 2016 and no one handling it properly to this day? I was rather excited than trying to be a jerk, but it was frustrating that it's lasted this long when so much impressive work gets done in other emulators and within RetroArch itself. (Problem has probably existed for years prior to that too)

this is no-ones job. this is voluntary. pcsx rearmed has had no real development for ~5 years since the original developer moved on, so is a bit of dead-end, and would be essentially redundant if beetle gets its long-awaited ARM dynarec, so be grateful this is getting any attention!

but never mind that: don't be a jerk for any reason is a good rule!

**Comment by jayare5** [comment]
Yikes. that would explain it. Now that the problem is better understood, is there anything else to be done in hopes for it getting fixed by voluntary work before we let it wait? Would this thread be enough or would another thread need to be made or anything?

**Comment by notaz** [comment]
There are now some options for this, including using the game's centering configuration. Not sure this is still relevant so closing for now.
